### ISSUE TITLE
Memory leak using UInt8Array

### ISSUE THREAD

**Issue by simontom** [microsoft/ClearScript#370]
Hey guys,
would u mind looking at this sky high rocketing unmanaged memory allocations which are not getting GC'ed?

We found this is happening since v7.1.5 (GC works like a charm in v7.1.4).
Happening on the latest version, too.

Repro ([according to this issue](https://github.com/microsoft/ClearScript/issues/126#issuecomment-505423728)):
```
using Microsoft.ClearScript.JavaScript;
using Microsoft.ClearScript.V8;

namespace V8EngineWorkbench;

internal static class Program
{
    private const int Size = 4096;
    private static readonly byte[] Bytes = CreateBytes();

    private static void Main()
    {
        for (var repetitions = 0; repetitions < 16; repetitions++)
        {
            CreateLeak();
        }
    }

    private static void CreateLeak()
    {
        var engine = new V8ScriptEngine();

        dynamic uint8ArrayFactoryFunc = engine.Evaluate(@"
            (function (length) {
                return new Uint8Array(length);
            })
            .valueOf();
        ");

        ITypedArray<byte> Alloc(long byteCount) => uint8ArrayFactoryFunc(byteCount);

        for (var i = 0; i < 10_000; i++)
        {
            var typedArray = Alloc(Size);
            typedArray.Write(Bytes, 0, Size, 0);
        }

        // engine.CollectGarbage(true); // Not helping
        engine.Dispose();
    }

    private static void CreateLeak2()
    {
        var engine = new V8ScriptEngine();

        engine.AddHostType(typeof(Task));

        engine.Execute(@"
            (async function () {
                for (let i = 0; i < 160_000; i++) {
                    const newData = new Uint8Array(4096);
                    newData.fill(42);

                    await Task.Delay(1);
                }
            })();
        ");

        engine.Dispose();
    }

    private static byte[] CreateBytes()
    {
        var bytes = new byte[Size];
        Array.Fill(bytes, (byte)42);
        return bytes;
    }
}
```
If u look at the _CreateLeak2_, all the data stays in the Engine and the result is the same.

Process memory profile from dotMemory on v7.1.5:
![image](https://user-images.githubusercontent.com/1636523/164723051-79a325c9-2d73-46fb-b633-9d09c2523cfa.png)

Process memory profile from dotMemory on v7.1.4:
![image](https://user-images.githubusercontent.com/1636523/164723362-b821e8c4-c0ae-40e7-9f35-6a50cd3e1667.png)

Tested on:
- ClearScript v7.1.4 (this one works pretty fine), v7.1.5, v7.2.3, v7.2.4
- .Net 6

**Comment by simontom** [comment]
Played w/ it a bit and found another funny thing.

This works pretty fine:
```
private static void CreateLeak()
    {
        var engine = new V8ScriptEngine();

        dynamic uint8ArrayFactoryFunc = engine.Evaluate(@"
            (function (length) {
                return new Uint8Array(length);
            })
            .valueOf();
        ");

        ITypedArray<byte> Alloc(long byteCount) => uint8ArrayFactoryFunc(byteCount);

        for (var i = 0; i < 10_000; i++)
        {
            var typedArray = Alloc(Size);
            typedArray.Write(Bytes, 0, Size, 0);

            if (i > 0 && i % 1000 == 0)
            {
                engine.CollectGarbage(true);
            }
        }

        // engine.CollectGarbage(true); // Not helping
        engine.Dispose();
    }
```

Still works pretty well (gets GC'ed at 0 and 5000):
```
    private static void CreateLeak()
    {
        var engine = new V8ScriptEngine();

        dynamic uint8ArrayFactoryFunc = engine.Evaluate(@"
            (function (length) {
                return new Uint8Array(length);
            })
            .valueOf();
        ");

        ITypedArray<byte> Alloc(long byteCount) => uint8ArrayFactoryFunc(byteCount);

        for (var i = 0; i < 10_000; i++)
        {
            var typedArray = Alloc(Size);
            typedArray.Write(Bytes, 0, Size, 0);

            if (i % 5_000 == 0)
            {
                engine.CollectGarbage(true);
            }
        }

        // engine.CollectGarbage(true); // Not helping
        engine.Dispose();
    }
```

However, if I use another indexes (e.g. 4 and 4000), it tries to reach the sky like Buzz Lightyear:
```
    private static void CreateLeak()
    {
        var engine = new V8ScriptEngine();

        dynamic uint8ArrayFactoryFunc = engine.Evaluate(@"
            (function (length) {
                return new Uint8Array(length);
            })
            .valueOf();
        ");

        ITypedArray<byte> Alloc(long byteCount) => uint8ArrayFactoryFunc(byteCount);

        for (var i = 0; i < 10_000; i++)
        {
            var typedArray = Alloc(Size);
            typedArray.Write(Bytes, 0, Size, 0);

            if (i is 4 or 4_000)
            {
                engine.CollectGarbage(true);
            }
        }

        // engine.CollectGarbage(true); // Not helping
        engine.Dispose();
    }
```

**Comment by ClearScriptLib** [comment]
Hi @simontom,

We've reproduced the issue. Thanks for reporting it.

Unfortunately, our testing indicates that V8 array buffer cleanup is a _mess_ at the moment. That may have to do with the V8 team's recent efforts to move increasing amounts of processing into the background.

In any case, V8 simply doesn't clean up array buffers consistently, and this particular scenario, with its multitude of short-lived V8 runtimes, appears to be a pathological case. As you've noted, forcing GC periodically is _sometimes_ effective, but it isn't a reliable solution.

We'll continue investigating this issue and searching for a fix or workaround. In the meantime, try disabling V8's background activities by setting [`V8GlobalFlags.DisableBackgroundWork`](https://microsoft.github.io/ClearScript/Reference/html/T_Microsoft_ClearScript_V8_V8GlobalFlags.htm).

Thanks again!

**Comment by ClearScriptLib** [comment]
Hi again,

Additional testing has revealed that a bug in ClearScript's support for [`MaxArrayBufferAllocation`](https://microsoft.github.io/ClearScript/Reference/html/P_Microsoft_ClearScript_V8_V8RuntimeConstraints_MaxArrayBufferAllocation.htm) is the root cause of this issue.

We'll include a fix in the next release. In the meantime, disabling V8's background activities (see above) should work.

Cheers!

**Comment by simontom** [comment]
Hello there, thanks for ur help.
1) Do you have anything on your mind you think this change might break?
2) Do you know the expected date of the new release?

**Comment by ClearScriptLib** [comment]
Hi @simontom,

>Do you have anything on your mind you think this change might break?

Disabling V8's background work shouldn't break anything, but it might have a negative impact on performance. In our experience the impact is slight, but it depends on the application.

>Do you know the expected date of the new release?

A new stable version of V8 is expected tomorrow. Barring unexpected issues, ClearScript 7.2.5 should be released about a week after that. Unfortunately, we can't guarantee release dates.

Thank you!

**Comment by ClearScriptLib** [comment]
Fixed in [Version 7.2.5](https://github.com/microsoft/ClearScript/releases/tag/7.2.5).
